File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/model/ConnectionModels.kt
package dev.bota.sdk.model

public data class DeviceConnectionSettings(
    public val enabledConnections: EnabledConnections,
    public val heartbeatEnabledConnections: EnabledConnections = enabledConnections,
    public val heartbeatUnknownMask: UByte = 0u,
    public val uploadNetworkPreference: List<ConnectionType>,
    public val powerManagement: PowerManagement = PowerManagement(),
    public val streamingEnabled: Boolean = true,
    public val streamingFlushIntervalSeconds: Int = 60,
) {
    public data class EnabledConnections(
        public val wifi: Boolean,
        public val cellular: Boolean,
    )

    public data class PowerManagement(
        public val wifiIdleTimeoutSeconds: Int = 180,
        public val cellularIdleTimeoutSeconds: Int = 180,
    )

    public sealed interface ConnectionType {
        public data object Wifi : ConnectionType
        public data object Ble : ConnectionType
        public data object Cellular : ConnectionType
        public data class Unknown(public val rawValue: UByte) : ConnectionType

        public companion object
    }

    public fun normalized(model: DeviceType): DeviceConnectionSettings {
        if (model != DeviceType.BotaNote) return this
        return copy(
            enabledConnections = enabledConnections.copy(cellular = false),
            heartbeatEnabledConnections = heartbeatEnabledConnections.copy(cellular = false),
            uploadNetworkPreference = uploadNetworkPreference.filterNot { it == ConnectionType.Cellular },
        )
    }
}

public data class ParsedConnectionSettings(
    public val settings: DeviceConnectionSettings,
    public val supportedVersion: Boolean,
)

public sealed interface WiFiConfigResult {
    public data object Success : WiFiConfigResult
    public data object InvalidGrant : WiFiConfigResult
    public data object GrantExpired : WiFiConfigResult
    public data object DecryptionError : WiFiConfigResult
    public data object StorageError : WiFiConfigResult
    public data class Unknown(public val rawValue: UByte) : WiFiConfigResult
}

public sealed interface WiFiConnectionStatus {
    public data object Idle : WiFiConnectionStatus
    public data object Connecting : WiFiConnectionStatus
    public data object Connected : WiFiConnectionStatus
    public data object Failed : WiFiConnectionStatus
    public data object Disconnected : WiFiConnectionStatus
    public data class Unknown(public val rawValue: UByte) : WiFiConnectionStatus
}

public data class WiFiStatusInfo(
    public val status: WiFiConnectionStatus,
    public val signalStrength: UByte? = null,
    public val ssid: String? = null,
    public val lastError: String? = null,
)

public data class WiFiScanNetwork(
    public val ssid: String,
    public val quality: UByte,
    public val isCurrent: Boolean,
    public val isOpen: Boolean,
)

public data class DeviceWiFiScanResult(
    public val networks: List<WiFiScanNetwork>,
    public val currentSsid: String?,
)

internal sealed interface WiFiScanUpdate {
    data class Pending(val rawStatus: UByte) : WiFiScanUpdate
    data class Done(val result: DeviceWiFiScanResult) : WiFiScanUpdate
}

File: platforms/apple/Sources/BotaAppleSDK/Models/ConnectionModels.swift
public enum ConnectionType: Equatable, Sendable {
    case wifi
    case ble
    case cellular
    case unknown(UInt8)
}

public struct DeviceConnectionSettings: Equatable, Sendable {
    public struct EnabledConnections: Equatable, Sendable {
        public var wifi: Bool
        public var cellular: Bool

        public init(wifi: Bool, cellular: Bool) {
            self.wifi = wifi
            self.cellular = cellular
        }
    }

    public struct PowerManagement: Equatable, Sendable {
        public var wifiIdleTimeoutSeconds: Int
        public var cellularIdleTimeoutSeconds: Int

        public init(wifiIdleTimeoutSeconds: Int, cellularIdleTimeoutSeconds: Int) {
            self.wifiIdleTimeoutSeconds = wifiIdleTimeoutSeconds
            self.cellularIdleTimeoutSeconds = cellularIdleTimeoutSeconds
        }
    }

    public var enabledConnections: EnabledConnections
    public var heartbeatEnabledConnections: EnabledConnections
    public var heartbeatUnknownMask: UInt8
    public var uploadNetworkPreference: [ConnectionType]
    public var powerManagement: PowerManagement
    public var streamingEnabled: Bool
    public var streamingFlushIntervalSeconds: Int

    public init(
        enabledConnections: EnabledConnections,
        heartbeatEnabledConnections: EnabledConnections? = nil,
        heartbeatUnknownMask: UInt8 = 0,
        uploadNetworkPreference: [ConnectionType],
        powerManagement: PowerManagement = .init(
            wifiIdleTimeoutSeconds: 180,
            cellularIdleTimeoutSeconds: 180
        ),
        streamingEnabled: Bool = true,
        streamingFlushIntervalSeconds: Int = 60
    ) {
        self.enabledConnections = enabledConnections
        self.heartbeatEnabledConnections = heartbeatEnabledConnections ?? enabledConnections
        self.heartbeatUnknownMask = heartbeatUnknownMask
        self.uploadNetworkPreference = uploadNetworkPreference
        self.powerManagement = powerManagement
        self.streamingEnabled = streamingEnabled
        self.streamingFlushIntervalSeconds = streamingFlushIntervalSeconds
    }

    public func normalized(for model: DeviceType) -> Self {
        guard model == .botaNote else { return self }
        var normalized = self
        normalized.enabledConnections.cellular = false
        normalized.heartbeatEnabledConnections.cellular = false
        normalized.uploadNetworkPreference.removeAll { $0 == .cellular }
        return normalized
    }
}

public struct ParsedConnectionSettings: Equatable, Sendable {
    public let settings: DeviceConnectionSettings
    public let supportedVersion: Bool

    public init(settings: DeviceConnectionSettings, supportedVersion: Bool) {
        self.settings = settings
        self.supportedVersion = supportedVersion
    }
}

public enum WiFiConfigResult: Equatable, Sendable {
    case success
    case invalidGrant
    case grantExpired
    case decryptionError
    case storageError
    case unknown(UInt8)
}

public enum WiFiConnectionStatus: Equatable, Sendable {
    case idle
    case connecting
    case connected
    case failed
    case disconnected
    case unknown(UInt8)
}

public struct WiFiStatusInfo: Equatable, Sendable {
    public var status: WiFiConnectionStatus
    public var signalStrength: UInt8?
    public var ssid: String?
    public var lastError: String?

    public init(
        status: WiFiConnectionStatus,
        signalStrength: UInt8? = nil,
        ssid: String? = nil,
        lastError: String? = nil
    ) {
        self.status = status
        self.signalStrength = signalStrength
        self.ssid = ssid
        self.lastError = lastError
    }
}

public struct WiFiScanNetwork: Equatable, Sendable {
    public var ssid: String
    public var quality: UInt8
    public var isCurrent: Bool
    public var isOpen: Bool

    public init(ssid: String, quality: UInt8, isCurrent: Bool, isOpen: Bool) {
        self.ssid = ssid
        self.quality = quality
        self.isCurrent = isCurrent
        self.isOpen = isOpen
    }
}

public struct DeviceWiFiScanResult: Equatable, Sendable {
    public var networks: [WiFiScanNetwork]
    public var currentSSID: String?

    public init(networks: [WiFiScanNetwork], currentSSID: String?) {
        self.networks = networks
        self.currentSSID = currentSSID
    }
}

enum WiFiScanUpdate: Equatable, Sendable {
    case pending(UInt8)
    case done(DeviceWiFiScanResult)
}
